service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/vm/LzsRetry.java

```diff
@@ -0,0 +1,153 @@
+  // an errorCode 0. This isn't a real HTTP status code, but we can check for it anyway.
+  private static final int TIMEOUT_STATUS_CODE = 0;
+
+  private final Instant operationTimeout;
+
+  // How long to wait between retries.
+  private Duration retryDuration;
+
+  LzsRetry() {
+    this.operationTimeout = now().plus(OPERATION_TIMEOUT);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  protected LzsRetry(Duration timeout) {
+    this.operationTimeout = now().plus(timeout);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @FunctionalInterface
+  public interface LzsVoidFunction {
+    void apply() throws ApiException, InterruptedException;
+  }
+
+  @FunctionalInterface
+  public interface LzsFunction<R> {
+    R apply() throws ApiException, InterruptedException;
+  }
+
+  /**
+   * Requests made through the LZS client library sometimes fail with timeouts, generally due to
+   * transient network or connection issues. When this happens, the client library will throw an API
+   * exceptions with status code 0 wrapping a SocketTimeoutException. These errors should always be
+   * retried.
+   */
+  public static boolean isTimeoutException(ApiException apiException) {
+    return apiException.getCode() == TIMEOUT_STATUS_CODE
+        && apiException.getCause() instanceof SocketTimeoutException;
+  }
+
+  public static <T> T retry(LzsRetry.LzsFunction<T> function)
+      throws ApiException, InterruptedException {
+    LzsRetry LzsRetry = new LzsRetry();
+    return LzsRetry.perform(function);
+  }
+
+  public static <T> T retry(LzsRetry.LzsFunction<T> function, Duration timeout)
+      throws ApiException, InterruptedException {
+    LzsRetry LzsRetry = new LzsRetry(timeout);
+    return LzsRetry.perform(function);
+  }
+
+  public static void retry(LzsRetry.LzsVoidFunction function)
+      throws ApiException, InterruptedException {
+    LzsRetry LzsRetry = new LzsRetry();
+    LzsRetry.performVoid(function);
+  }
+
+  public static void retry(LzsRetry.LzsVoidFunction function, Duration timeout)
+      throws ApiException, InterruptedException {
+    LzsRetry LzsRetry = new LzsRetry(timeout);
+    LzsRetry.performVoid(function);
+  }
+
+  private <T> T perform(LzsRetry.LzsFunction<T> function)
+      throws ApiException, InterruptedException {
+    while (true) {
+      try {
+        return function.apply();
+      } catch (ApiException ex) {
+        if (isRetryable(ex)) {
+          logger.info("LzsRetry: caught retry-able exception: ", ex);
+          sleepOrTimeoutBeforeRetrying(ex);
+        } else {
+          throw ex;
+        }
+      } catch (ProcessingException ws) {
+        logger.info("LzsRetry: caught retry-able ProcessingException: ", ws);
+        sleepOrTimeoutBeforeRetrying(ws);
+      }
+    }
+  }
+
+  private boolean isRetryable(ApiException apiException) {
+    return isTimeoutException(apiException)
+        || apiException.getCode() == HttpStatus.SC_FORBIDDEN
```

Comment:
This is problematic in my testing; lzs returns an (expected?) 403 during the poll for a landing zone during creation, which causes gateway timeouts as this goes into an extensive retry. I will need to think about this a bit.

Reply:
Should 403 be a retry-able status code? I guess if we're seeing intermittent 403s that eventually go away then we might need it to be, but I'm generally not inclined to retry 403s as long as we trust LZS and Sam to return 403s correctly.

Reply:
Yeah, I think I agree. This was originally installed in WSM for TPS calls as we still get occasional spurious 403s from TPS from the oidc proxy. Might not be worth it in this case.  (see also my investigations into unknown error -102 AH01095)